### PR TITLE
Fixed Gitlab event collector fetch

### PR DESCRIPTION
**Pull request by yaakovpraisler**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/XSUP-28386)

## Description
Fixed the GitLab event collector fetch.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 1]
```diff
@@ -1,179 +1,285 @@
...
+           and returns the events that have been accumulated so far.
+
+        3. Reaches a rate limit.
+           In this case, stores the last URL suffix used in the lastRun obj
+           and returns the events that have been accumulated so far.
+        """
+        aggregated_events: List[dict] = []
+
+        user_defined_limit = user_defined_params.get('limit') or DEFAULT_LIMIT
+        current_url = params_url_suffix
```
Comment (dantavori) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:86]:
```suggestion
        current_url: str | None = params_url_suffix
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 2]
```diff
@@ -1,179 +1,285 @@
...
-            events = response.json()
-            events.sort(key=lambda k: k.get('created_at'))
-            if not events:
+            events, next_url = self.handle_pagination_first_batch(current_url, last_run)
+            while events:
+                for event in events:
+
+                    if event.get('id') == last_run.get('last_id'):
+                        demisto.debug('Encountered an event that was already fetched - stopping.')
+                        current_url = None  # type:ignore[assignment]
```
Comment (dantavori) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:95]:
```suggestion
                        current_url = None
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 3]
```diff
@@ -1,179 +1,285 @@
...
+                    # Finished iterating through all events in this batch
+                    if next_url:
+                        demisto.debug('Using the given next_url from the last API call to execute the next call.')
+                        current_url = next_url.split('?', 1)[1]
+                        events, next_url = self.get_events_request(current_url)
+                        continue
+                    else:
+                        current_url = None  # type:ignore[assignment]
+
+                demisto.debug('Finished iterating through all events in this fetch run.')
```
Comment (dantavori) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:117]: "... for the current event type"

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 4]
```diff
@@ -1,179 +1,285 @@
...
 
-    get_events = GetEvents(client, options)
+    query_params_url_suffix = prepare_query_params(params, last_run["audit_events"])
+    audit_events = client.fetch_events(query_params_url_suffix, last_run["audit_events"], params)
+    demisto.debug(f"Aggregated audits events: {len(audit_events)}")
+
+    group_and_project_events = []
+    for event_type in ['groups', 'projects']:
+        for obj_id in events_types_ids.get(f'{event_type}_ids', []):
+            client._base_url = f"{params['url']}/api/v4/{event_type}/{obj_id}/audit_events"
```
Comment (dantavori) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:212]: must we change _base_url? best practice is to leave it as-is, only the suffix should be sent

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 5]
```diff
@@ -1,179 +1,285 @@
...
+''' CLIENT CLASS '''
+
+
+class Client(BaseClient):
+
+    def get_events_request(self, next_url_link: str) -> tuple:
+        """
+        Sends request to get events from GitLab.
+        """
+        response = self._http_request(method='GET', url_suffix=f'?{next_url_link}', resp_type='response')
```
Comment (dantavori) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:41]:
```suggestion
    def get_events_request(self, url_suffix: str, query_params: str) -> tuple:
        """
        Sends request to get events from GitLab.
        """
        response = self._http_request(method='GET', url_suffix=f'{url_suffix}?{query_params}', resp_type='response')
```
added url_suffix as suggested in the previous comment, lmk if I am missing somthing

File: Packs/GitLab/ReleaseNotes/2_2_19.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+
+##### GitLab Event Collector
+
+- Fixed an issue where the ***fetch-events*** failed in case there are a large number of events and it didn't had enough time to fetch everything before it raised on timeout.
```
Comment (dantavori) [inline Packs/GitLab/ReleaseNotes/2_2_19.md:6]:
```suggestion
- Fixed an issue where the ***fetch-events*** command failed on timeout.
```

**Review by dantavori — COMMENTED** [review]
very nice, see notes

**Review by dantavori — APPROVED** [review]

**Comment by yaakovpraisler** [comment]
Failes on `Unit-tests for 'Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py' must reach a coverage of at least 99.0% (currently: 86.89%).`
@dantavori - need to force merge here.

**Pull request merged**